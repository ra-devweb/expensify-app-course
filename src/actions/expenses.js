import database from '../firebase/firebase';

// Add expense
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense,
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0,
    } = expenseData;

    const expense = { description, note, amount, createdAt };

    return database
      .ref('expenses')
      .push(expense)
      .then((ref) => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense,
          })
        );
      });
  };
};

// Delete expense
export const deleteExpense = ({ id } = {}) => ({ type: 'DELETE_EXPENSE', id });

// Update expense
export const updateExpense = (id, update) => ({
  type: 'UPDATE_EXPENSE',
  id,
  update,
});

// Set expenses
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses,
});

export const startSetExpenses = () => {
  return (dispatch) => {
    return database
      .ref('expenses')
      .once('value')
      .then((snapshot) => {
        const expenses = [];
        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setExpenses(expenses));
      });
  };
};
