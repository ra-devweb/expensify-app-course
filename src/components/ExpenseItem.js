import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

numeral.register('locale', 'dr', {
  delimiters: {
    thousands: ' ',
    decimal: ',',
  },
  abbreviations: {
    thousand: 'k',
    million: 'm',
    billion: 'b',
    trillion: 't',
  },
  ordinal: function (number) {
    return number === 1 ? 'er' : 'Ã¨me';
  },
  currency: {
    symbol: 'MAD',
  },
});

numeral.locale('dr');

const ExpenseItem = ({ id, description, amount, createdAt }) => {
  return (
    <div>
      <Link to={`/edit/${id}`}>
        <h1>{description}</h1>
      </Link>
      <p>{numeral(amount / 100).format('0,0.00 $')}</p>
      <span>{moment(createdAt).format('MMMM Do, YYYY')}</span>
    </div>
  );
};

export default ExpenseItem;
