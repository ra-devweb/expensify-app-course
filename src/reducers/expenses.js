// Expense default state param
const expensifyDefaultState = [];

// Expensify reducer
export default (state = expensifyDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense];
    case 'DELETE_EXPENSE':
      return state.filter(({ id }) => id !== action.id);
    case 'UPDATE_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.update,
          };
        } else {
          return expense;
        }
      });
    default:
      return state;
  }
};
