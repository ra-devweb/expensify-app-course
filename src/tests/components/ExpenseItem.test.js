import React from 'react';
import { shallow } from 'enzyme';

import ExpenseItem from '../../components/ExpenseItem';
import expenses from '../fixtures/expenses';

test('Should render expense item component correctly', () => {
  const wrapper = shallow(<ExpenseItem {...expenses[0]} />);

  expect(wrapper).toMatchSnapshot();
});
