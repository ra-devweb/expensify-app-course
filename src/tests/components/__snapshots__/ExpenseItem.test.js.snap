// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expense item component correctly 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h1>
      Rent
    </h1>
  </Link>
  <p>
    35,00 MAD
  </p>
  <span>
    January 1st, 1970
  </span>
</div>
`;
